Several possible problems within the ZPublisher Converters module

When working on #557 I spotted several possible problems within the Converters module.

The converters which inherit from `_unicode_converter` all fail under Python 3 and byte input.

Testcase:
```
def test_field2utext_with_binary_string(self):
    from ZPublisher.Converters import field2utext
    to_convert = b'abc'
    if PY2:
        expected = u'abc'
        self.assertEqual(field2utext(to_convert), expected)
    if PY3:
        expected = 'abc'
        self.assertEqual(field2utext(to_convert), expected)

Failure in test test_field2utext_with_binary_string (ZPublisher.tests.test_Converters.ConvertersTests)
Traceback (most recent call last):
  File "/usr/lib64/python3.6/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.6/unittest/case.py", line 622, in run
    testMethod()
  File "/home/jugmac00/Projects/Zope/Zope/src/ZPublisher/tests/test_Converters.py", line 259, in test_field2utext_with_binary_string
    self.assertEqual(field2utext(to_convert), expected)
  File "/usr/lib64/python3.6/unittest/case.py", line 846, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.6/unittest/case.py", line 1220, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib64/python3.6/unittest/case.py", line 687, in fail
    raise self.failureException(msg)
AssertionError: "b'abc'" != 'abc'
- b'abc'
+ abc
```


This problem applies to field2utokens, field2utext, field2string. field2ulines on the other hand does not inherit from _unicode_converter and has to be examined separately.

I spotted https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L186 as the root problem for this behaviour, at least with Python 3.

----

Another possible problem is in https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L209 - as line breaks only get normalized when there is input of type byte.


I tried to fix these "problems" yesterday - seemingly successfully, but then suddenly https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/tests/testHTTPRequest.py#L302 failed for the utext case.

----

Also, can anybody shed some light on what this "read" attribute is all about?
e.g. https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L221


Before spending more time on these issue, I'd like somebody to confirm that these are real issues.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -20,6 +20,9 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
   ``rfc850_date``, and ``rfc1123_date`` which used to be in ``App.Common``
   keeping backwards-compatibility imports in place.
 
+- With the exception of ``field2bytes``, field converters should not handle
+  files (`#558 <https://github.com/zopefoundation/Zope/issues/558>`_)
+
 
 5.1.2 (2021-03-02)
 ------------------

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -8,8 +8,8 @@ The change log for the previous version, Zope 4, is at
 https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
 
 
-5.1.3 (unreleased)
-------------------
+5.2 (unreleased)
+----------------
 
 - Updated/fixed the poll application tutorial in the Zope Developers Guide
   (`#958 <https://github.com/zopefoundation/Zope/issues/958>`_)
@@ -20,8 +20,12 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
   ``rfc850_date``, and ``rfc1123_date`` which used to be in ``App.Common``
   keeping backwards-compatibility imports in place.
 
-- With the exception of ``field2bytes``, field converters should not handle
-  files (`#558 <https://github.com/zopefoundation/Zope/issues/558>`_)
+Backwards incompatible changes
+++++++++++++++++++++++++++++++
+
+- With the exception of ``field2bytes``, field converters do no longer try to
+  read file like objects
+  (`#558 <https://github.com/zopefoundation/Zope/issues/558>`_)
 
 
 5.1.2 (2021-03-02)

--- a/setup.py
+++ b/setup.py
@@ -29,7 +29,7 @@ def _read_file(filename):
 README = _read_file('README.rst')
 CHANGES = _read_file('CHANGES.rst')
 
-version = '5.1.3.dev0'
+version = '5.2.dev0'
 
 
 setup(

--- a/src/ZPublisher/Converters.py
+++ b/src/ZPublisher/Converters.py
@@ -23,13 +23,8 @@ default_encoding = 'utf-8'
 
 
 def field2string(v):
-    """Converts value to native strings.
-
-    So always to `str` no matter which Python version you are on.
-    """
-    if hasattr(v, 'read'):
-        return v.read()
-    elif isinstance(v, bytes):
+    """Converts value to string."""
+    if isinstance(v, bytes):
         return v.decode(default_encoding)
     else:
         return str(v)

--- a/src/ZPublisher/Converters.py
+++ b/src/ZPublisher/Converters.py
@@ -167,8 +167,6 @@ class _unicode_converter:
         #       <input name="description:utf8:ustring" .....
         # rather than
         #       <input name="description:ustring" .....
-        if hasattr(v, 'read'):
-            v = v.read()
         v = str(v)
         return self.convert_unicode(v)
 

--- a/src/ZPublisher/Converters.py
+++ b/src/ZPublisher/Converters.py
@@ -204,8 +204,6 @@ field2utext = field2utext()
 
 class field2ulines:
     def __call__(self, v):
-        if hasattr(v, 'read'):
-            v = v.read()
         if isinstance(v, (list, tuple)):
             return [field2ustring(x) for x in v]
         v = str(v)

--- a/src/ZPublisher/Converters.py
+++ b/src/ZPublisher/Converters.py
@@ -31,7 +31,7 @@ def field2string(v):
 
 
 def field2bytes(v):
-    # Converts value to bytes.
+    """Converts value to bytes."""
     if hasattr(v, 'read'):
         return v.read()
     elif isinstance(v, str):

--- a/src/ZPublisher/tests/test_Converters.py
+++ b/src/ZPublisher/tests/test_Converters.py
@@ -160,6 +160,14 @@ class ConvertersTests(unittest.TestCase):
         expected = b'to_convert'
         self.assertEqual(field2bytes(to_convert), expected)
 
+    def test_field2bytes_with_filelike_object(self):
+        from ZPublisher.Converters import field2bytes
+        to_convert = b'to_convert'
+        class Filelike:
+            def read(self):
+                return to_convert
+        self.assertEqual(field2bytes(Filelike()), to_convert)
+
     def test_field2date_international_with_proper_date_string(self):
         from ZPublisher.Converters import field2date_international
         to_convert = "2.1.2019"

--- a/src/ZPublisher/tests/test_Converters.py
+++ b/src/ZPublisher/tests/test_Converters.py
@@ -149,15 +149,6 @@ class ConvertersTests(unittest.TestCase):
         expected = 'to_convert'
         self.assertEqual(field2string(to_convert), expected)
 
-    def test_field2string_with_filelike_object(self):
-        from ZPublisher.Converters import field2string
-        to_convert = 'to_convert'
-
-        class Filelike:
-            def read(self):
-                return to_convert
-        self.assertEqual(field2string(Filelike()), to_convert)
-
     def test_field2bytes_with_bytes(self):
         from ZPublisher.Converters import field2bytes
         to_convert = b'to_convert'

--- a/src/ZPublisher/tests/test_Converters.py
+++ b/src/ZPublisher/tests/test_Converters.py
@@ -163,6 +163,7 @@ class ConvertersTests(unittest.TestCase):
     def test_field2bytes_with_filelike_object(self):
         from ZPublisher.Converters import field2bytes
         to_convert = b'to_convert'
+
         class Filelike:
             def read(self):
                 return to_convert

